File: APIApplication/src/main/java/org/apiapplication/services/interfaces/SubjectService.java
package org.apiapplication.services.interfaces;

import org.apiapplication.dto.common.IdDto;
import org.apiapplication.dto.subject.AddSubjectDto;
import org.apiapplication.dto.subject.SubjectDto;
import org.apiapplication.dto.subject.UpdateSubjectDto;

import java.util.List;

public interface SubjectService {
    SubjectDto getSubjectById(int id);

    List<SubjectDto> get(Integer universityId);

    IdDto add(AddSubjectDto addSubjectDto);

    void update(int subjectId, UpdateSubjectDto updateSubjectDto);

    void delete(int subjectId);
}

File: FrontendApplication/frontend-project/src/app/shared/services/subject.service.ts
import { HttpClient, HttpParams } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { baseUrl } from "../constants/url.constant";
import { IdDto } from "../models/common.model";
import { AddSubjectDto, UpdateSubjectDto, SubjectDto } from "../models/subject.model";

@Injectable({
    providedIn: 'root'
})
export class SubjectsService {
    private url: string = baseUrl + "/subjects";

    constructor(private readonly httpClient: HttpClient) {
    }

    getAll(): Observable<SubjectDto[]> {
        return this.httpClient.get<SubjectDto[]>(`${this.url}`);
    }

    getById(id: number): Observable<SubjectDto> {
        return this.httpClient.get<SubjectDto>(`${this.url}/${id}`);
    }

    add(dto: AddSubjectDto): Observable<IdDto> {
        return this.httpClient.post<IdDto>(`${this.url}`, dto);
    }

    update(id: number, dto: UpdateSubjectDto): Observable<any> {
        return this.httpClient.put<any>(`${this.url}/${id}`, dto);
    }

    delete(id: number): Observable<any> {
        const queryParams = new HttpParams();
        queryParams.set('id', id);

        return this.httpClient.delete<any>(`${this.url}/${id}`, { params: queryParams });
    }
}